File: mandalay_bay/resort_bridge.py
"""Cross-resort requirement checks — room, pool, shopping, rewards tier."""

from __future__ import annotations

from mandalay_bay.casino_amenities import ensure_amenities
from mandalay_bay.hotel import ensure_hotel
from mandalay_bay.pool_complex import ensure_pool_complex
from mandalay_bay.rewards import tier_for_wagered
from mandalay_bay.rewards_perks import tier_index
from mandalay_bay.session import PlayerSession


def _has_all(haystack: list[str], needles: list[str]) -> bool:
    return all(n in haystack for n in needles)


def get_session_tier_index(session: PlayerSession) -> int:
    rewards = getattr(session, "rewards", None)
    wagered = rewards.lifetime_wagered if rewards else 0
    return tier_index(tier_for_wagered(wagered).id)


def resort_requirements_met(session: PlayerSession, req: dict, *, hotel=None, room_amenities=None) -> bool:
    if not req:
        return True
    hotel = hotel or ensure_hotel(session)
    ra = room_amenities or getattr(hotel, "room_amenities", None)
    pc = ensure_pool_complex(session)
    amenities = ensure_amenities(session)

    if "pool_zones" in req and not _has_all(pc.visited_zones, req["pool_zones"]):
        return False
    if "pool_events" in req and not _has_all(pc.unlocked_events, req["pool_events"]):
        return False
    if ra and "room_events" in req and not _has_all(ra.unlocked_events, req["room_events"]):
        return False
    if "shopping_items" in req and not _has_all(amenities.purchased_items, req["shopping_items"]):
        return False
    if "min_tier_index" in req and get_session_tier_index(session) < req["min_tier_index"]:
        return False
    if "room_types" in req and hotel.room_type not in req["room_types"]:
        return False
    return True

File: docs/js/resort-bridge.js
/** Cross-resort requirement checks — room, pool, shopping, rewards tier. */

import { ensureAmenities } from "./casino-amenities.js";
import { ensurePoolComplex } from "./pool-complex.js";
import { ensureHotel } from "./hotel.js";
import { tierForWagered } from "./rewards.js";
import { tierIndex } from "./rewards-perks.js";

function hasAll(haystack, needles) {
  return needles.every((n) => haystack.includes(n));
}

/** @param {import("./core.js").PlayerSession} session */
export function getSessionTierIndex(session) {
  const wagered = session.rewards?.lifetimeWagered ?? 0;
  return tierIndex(tierForWagered(wagered).id);
}

/** @param {import("./core.js").PlayerSession} session */
export function getPurchasedItemIds(session) {
  return ensureAmenities(session).purchasedItems ?? [];
}

/** @param {import("./core.js").PlayerSession} session */
export function getPoolUnlockedEvents(session) {
  return ensurePoolComplex(session).unlockedEvents ?? [];
}

/** @param {import("./core.js").PlayerSession} session */
export function getPoolVisitedZones(session) {
  return ensurePoolComplex(session).visitedZones ?? [];
}

/**
 * Evaluate shared resort requirements used by room and pool events.
 * @param {import("./core.js").PlayerSession} session
 * @param {object} req
 * @param {object} [ctx]
 * @param {object} [ctx.hotel]
 * @param {object} [ctx.roomAmenities]
 */
export function resortRequirementsMet(session, req, ctx = {}) {
  if (!req) return true;
  const hotel = ctx.hotel ?? ensureHotel(session);
  const ra = ctx.roomAmenities ?? hotel.roomAmenities ?? {};
  const pc = ensurePoolComplex(session);
  const amenities = ensureAmenities(session);

  if (req.poolZones && !hasAll(pc.visitedZones ?? [], req.poolZones)) return false;
  if (req.poolEvents && !hasAll(pc.unlockedEvents ?? [], req.poolEvents)) return false;
  if (req.roomEvents && !hasAll(ra.unlockedEvents ?? [], req.roomEvents)) return false;
  if (req.shoppingItems && !hasAll(amenities.purchasedItems ?? [], req.shoppingItems)) return false;
  if (req.minTierIndex != null && getSessionTierIndex(session) < req.minTierIndex) return false;
  if (req.roomTypes && !req.roomTypes.includes(hotel.roomType)) return false;
  return true;
}

/** Cryptic hint text for locked room events (no exact recipes). */
export const EVENT_HINTS = {
  shark_whisperer: "Something about sharks and snacks…",
  midnight_ocean: "Tokyo at 3 AM meets the reef cam…",
  champagne_sunset: "Wave pool on TV, something bubbly in hand…",
  pool_party_vip: "Concierge, champagne, and maybe the pool deck…",
  what_happens: "Vodka, voicemail, and the Strip below…",
  high_roller_crawl: "Suite neighbors and something sparkling…",
  hangover_brunch: "Survive the night first…",
  buckingham_moment: "Dial a number you shouldn't have…",
  steve_harvey_hotline: "Steve on TV, Pete on the phone, net-positive on the floor…",
  fight_night_suite: "Arena replay, off-strip picks, suite view…",
  foundation_after_dark: "Noir status, penthouse, velvet rope…",
  sky_bridge_haul: "Mandalay Place shopping, balcony, spa recovery…",
  eleven_acres_hangover: "Conquer the pool complex, then brunch…",
  chapel_wrong_turn: "Wrong number, wedding bells…",
  convention_survival: "Keynote TV, energy drink, privacy sign…",
  butler_turn_down: "Penthouse only — let the butler in…",
  telescope_strip: "Penthouse balcony — point at the sportsbook…",
};

export function hintForEvent(eventId) {
  return EVENT_HINTS[eventId] ?? "Keep exploring the resort…";
}
